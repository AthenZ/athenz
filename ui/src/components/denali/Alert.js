/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';
import Icon from '../denali/icons/Icon';
import {
    colors,
    cssDropShadow,
    cssFontStyles,
    cssFontWeights,
} from '../denali/styles';
import { css, cx } from 'emotion';
import { rgba } from 'polished';

const TYPE_COLOR_MAP = {
    danger: 'statusDanger',
    info: 'statusInfo',
    success: 'statusSuccess',
    warning: 'statusWarning',
};

const TYPE_ICON_MAP = {
    danger: 'stop-warning',
    info: 'information-circle',
    success: 'check-circle',
    warning: 'warning',
};

// http://reactcommunity.org/react-modal/examples/set_app_element.html
ReactModal.setAppElement('body');

const modalOverlayClass = css`
    background-color: transparent;
    bottom: 0;
    left: 0;
    position: fixed;
    right: 0;
    top: 0;
    z-index: 5;
    label: modal-overlay;
`;

const makeAlertClass = (props) => css`
    ${cssDropShadow};
    align-items: flex-start;
    background: ${colors.white};
    border-radius: 4px;
    border-top: 4px solid ${colors[TYPE_COLOR_MAP[props.type]]};
    box-sizing: border-box;
    color: ${colors.black};
    display: flex;
    flex-flow: row nowrap;
    ${cssFontStyles.default};
    left: 50%;
    max-height: calc(100vh - 160px);
    min-width: 400px;
    outline: none;
    position: fixed;
    transform: translate(-50%, 0);
    width: 400px;
    label: alert;
    &.ReactModal__Content {
        opacity: 0;
        top: 0;
    }
    &.animated {
        transition: opacity 0.2s ease-out,
            top 0.3s cubic-bezier(0.175, 0.885, 0.32, 1.275);
    }
    &.ReactModal__Content--after-open {
        top: 80px;
        opacity: 1;
    }
    /* Animate fade out on close. This is currently not used as tests fail */
    &.ReactModal__Content--before-close {
        opacity: 0;
    }
    & .left-section {
        flex: 0 0 28px;
        padding: 14px;
        & .alert-icon {
            align-items: flex-start;
            display: flex;
            justify-content: center;
        }
    }
    & .right-section {
        display: flex;
        flex: 1 1 auto;
        flex-flow: column nowrap;
        & .title-row {
            align-items: center;
            display: flex;
            flex-flow: row nowrap;
            height: initial;
            justify-content: space-between;
            padding: 14px 14px 14px 0;
            & .title {
                ${cssFontWeights.bold};
            }
        }
        & .description {
            font-size: 14px;
            font-weight: 300;
            padding: 0 14px 14px 0;
        }
    }
`;

/**
 * Alert displays a dismissable modal near the top of the screen. It is meant to
 * show a short blurb of information. There are four
 * alert types:
 *
 * 1. Info
 * 2. Success
 * 3. Warning
 * 4. Danger
 *
 * Under the hood, Alert uses [`react-modal`](http://reactcommunity.org/react-modal/).
 * To prevent the page from scrolling when the modal is showing, add the
 * following to the app's global CSS:
 *
 * ```css
 * .ReactModal__Body--open {
 *   overflow: hidden;
 * }
 */
class Alert extends React.Component {
    // Taken from:
    // https://stackoverflow.com/questions/45678517/better-way-to-cleartimeout-in-componentwillunmount
    setTimer = (ms) => {
        // Remember the timer handle
        this.timerHandle = setTimeout(() => {
            this.timerHandle = null;
            this.props.onClose();
        }, ms);
    };

    clearTimer = () => {
        // Is our timer running?
        if (this.timerHandle) {
            clearTimeout(this.timerHandle);
            this.timerHandle = null;
        }
    };

    componentDidMount() {
        if (this.props.isOpen && this.props.duration) {
            this.setTimer(this.props.duration);
        }
    }

    componentDidUpdate(prevProps) {
        if (!prevProps.isOpen && this.props.isOpen && this.props.duration) {
            this.setTimer(this.props.duration);
        } else if (prevProps.isOpen && !this.props.isOpen) {
            this.clearTimer();
        }
    }

    componentWillUnmount() {
        this.clearTimer();
    }

    render() {
        const StatusIcon = TYPE_ICON_MAP[this.props.type];
        const alertClass = makeAlertClass(this.props);

        return (
            <ReactModal
                aria={{
                    labelledby: 'alert-title',
                    describedby: 'alert-description',
                }}
                className={cx(
                    alertClass,
                    'denali-alert',
                    `denali-alert-${this.props.type}`,
                    {
                        animated: !this.props.noanim,
                    },
                    this.props.className
                )}
                closeTimeoutMS={200}
                descriptionLabel={`${this.props.type} alert`}
                isOpen={this.props.isOpen}
                overlayClassName={modalOverlayClass}
                shouldCloseOnOverlayClick={false}
                onRequestClose={this.props.onClose}
            >
                <div className='left-section'>
                    <div className='alert-icon' data-testid='alert-icon'>
                        <Icon
                            icon={'check-circle'}
                            color={colors[TYPE_COLOR_MAP[this.props.type]]}
                            decorative={true}
                            size='28px'
                            verticalAlign='middle'
                        />
                    </div>
                </div>
                <div className='right-section'>
                    <div className='title-row'>
                        <div
                            className='title'
                            id='alert-title'
                            data-testid='alert-title'
                        >
                            {this.props.title}
                        </div>
                        <div
                            className='close-button'
                            onClick={this.props.onClose}
                        >
                            <Icon
                                color={rgba(colors.grey800, 0.4)}
                                colorHover={rgba(colors.grey800, 0.6)}
                                isLink
                                size='28px'
                                icon={'close'}
                            />
                        </div>
                    </div>
                    {this.props.description && (
                        <div
                            className='description'
                            id='alert-description'
                            data-testid='alert-description'
                        >
                            {this.props.description}
                        </div>
                    )}
                </div>
            </ReactModal>
        );
    }
}

Alert.propTypes = {
    /** Additonal class to apply to the outer div of the modal content */
    className: PropTypes.string,
    /** Optional description. Can be a string or JSX / component */
    description: PropTypes.any,
    /** Once open, call `onClose()` after N milliseconds */
    duration: PropTypes.number,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Toggle the alert on or off */
    isOpen: PropTypes.bool,
    /** Title of the alert */
    title: PropTypes.string.isRequired,
    /** Alert supports four status types */
    type: PropTypes.oneOf(['danger', 'info', 'success', 'warning']),
    /** Handler for closing the alert */
    onClose: PropTypes.func.isRequired,
};

Alert.defaultProps = {
    duration: 0,
    isOpen: false,
    noanim: false,
    type: 'info',
};

export default Alert;
