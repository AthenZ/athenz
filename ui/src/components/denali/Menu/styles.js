/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { css } from 'emotion';
import { cssFontStyles } from '../styles/fonts';
import { colors } from '../styles/colors';
import { cssDropShadow } from '../styles/drop-shadow';

export const makeCssPopperBox = (props) => css`
    ${cssDropShadow};
    ${cssFontStyles.default};
    background-color: ${colors.white};
    border: 2px solid ${colors.border};
    border-radius: 4px;
    color: ${colors.black};
    cursor: default;
    line-height: 1.4;
    margin: 0;
    max-width: ${props.basic ? '400px' : undefined};
    min-width: 20px;
    padding: ${props.basic ? '6px 10px' : 0};
    transition: ${!props.noanim && 'opacity 0.25s ease-in'};
    z-index: 999;
    label: denali-menu;
    /* Shift the popper to allow space for arrow */
    &[data-placement*='bottom'] {
        margin-top: 8px;
    }
    &[data-placement*='top'] {
        margin-bottom: 8px;
    }
    &[data-placement*='left'] {
        margin-right: 8px;
    }
    &[data-placement*='right'] {
        margin-left: 8px;
    }

    /**
   * Invisible padding between trigger and main content (around the arrow) so
   * user can slide from one to the other without the menu closing.
   */
    &::after {
        background: transparent;
        content: '';
        position: absolute;
    }
    &[data-placement*='top']::after {
        bottom: -12px;
        height: 12px;
        width: 100%;
    }
    &[data-placement*='bottom']::after {
        height: 12px;
        top: -12px;
        width: 100%;
    }
    &[data-placement*='left']::after {
        right: -12px;
        height: 100%;
        width: 12px;
    }
    &[data-placement*='right']::after {
        left: -12px;
        height: 100%;
        width: 12px;
    }
`;

export const cssArrow = css`
    height: 1rem;
    position: absolute;
    width: 1rem;
    label: menu-arrow;

    &::before,
    &::after {
        border-style: solid;
        content: '';
        display: block;
        height: 0;
        margin: auto;
        position: absolute;
        width: 0;
    }

    &[data-placement*='bottom'] {
        left: 0;
        margin-top: -6px;
        top: -8px;
        width: 1rem;
    }

    &[data-placement*='bottom']::before {
        border-color: transparent transparent ${colors.grey400} transparent;
        border-width: 0 8px 8px 8px;
        position: absolute;
        top: 4px;
    }

    &[data-placement*='bottom']::after {
        border-color: transparent transparent ${colors.white} transparent;
        border-width: 0 8px 8px 8px;
        top: 7px;
    }

    &[data-placement*='top'] {
        bottom: 0;
        height: 1rem;
        left: 0;
        margin-bottom: -1rem;
        width: 1rem;
    }

    &[data-placement*='top']::before {
        border-color: ${colors.grey400} transparent transparent transparent;
        border-width: 8px 8px 0 8px;
        position: absolute;
        top: 2px;
    }

    &[data-placement*='top']::after {
        border-color: ${colors.white} transparent transparent transparent;
        border-width: 8px 8px 0 8px;
        top: -1px;
    }

    &[data-placement*='right'] {
        height: 1rem;
        left: 0;
        margin-left: -1rem;
        width: 1rem;
    }

    &[data-placement*='right']::before {
        border-color: transparent ${colors.grey400} transparent transparent;
        border-width: 8px 8px 8px 0;
        left: 5px;
    }

    &[data-placement*='right']::after {
        border-color: transparent ${colors.white} transparent transparent;
        border-width: 8px 8px 8px 0;
        left: 8px;
        top: 0;
    }

    &[data-placement*='left'] {
        height: 1rem;
        margin-right: -1rem;
        right: 0;
        width: 1rem;
    }

    &[data-placement*='left']::before {
        border-color: transparent transparent transparent ${colors.grey400};
        border-width: 8px 0 8px 8px;
        left: 2px;
    }

    &[data-placement*='left']::after {
        border-color: transparent transparent transparent ${colors.white};
        border-width: 8px 0 8px 8px;
        left: -1px;
        top: 0;
    }
`;
