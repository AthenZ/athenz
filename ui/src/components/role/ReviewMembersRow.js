/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import styled from '@emotion/styled';
import RadioButton from '../denali/RadioButton';
import DateUtils from '../utils/DateUtils';

const TDStyled = styled.td`
    background-color: ${(props) => props.color};
    text-align: ${(props) => props.align};
    padding: 5px 0 5px 15px;
    vertical-align: middle;
    word-break: break-all;
    border-bottom: 1px solid #d5d5d5;
`;

const TrStyled = styled.tr`
    background-color: ${(props) => props.color};
`;

export default class ReviewMembersRow extends React.Component {
    constructor(props) {
        super(props);
        this.api = this.props.api;
        this.onReview = this.onReview.bind(this);
        this.state = {
            selectedOption: 'extend',
        };
        this.localDate = new DateUtils();
    }

    onReview(evt) {
        switch (evt.target.value) {
            case 'extend':
                this.props.onUpdate('extend', this.props.details.memberName);
                break;
            case 'delete':
                this.props.onUpdate('delete', this.props.details.memberName);
                break;
            case 'no-action':
                this.props.onUpdate('no-action', this.props.details.memberName);
                break;
            default:
                break;
        }
        this.setState({
            selectedOption: evt.target.value,
        });
    }

    componentDidUpdate = (prevProps) => {
        if (prevProps.submittedReview !== this.props.submittedReview) {
            this.setState({
                selectedOption: 'extend',
            });
        }
    };

    render() {
        let rows = [];
        let left = 'left';
        let center = 'center';
        let member = this.props.details;
        let color = this.props.color;
        let exp = member.expiration;
        if (exp) {
            exp = this.localDate.getLocalDate(exp, 'UTC', 'UTC');
        }

        rows.push(
            <TrStyled key={this.props.idx} data-testid='role-review-row'>
                <TDStyled color={color} align={left}>
                    {member.memberName}
                </TDStyled>
                <TDStyled color={color} align={left}>
                    {member.memberFullName}
                </TDStyled>
                <TDStyled color={color} align={left}>
                    {exp}
                </TDStyled>
                <TDStyled color={color} align={center}>
                    <RadioButton
                        name={this.props.role + this.props.idx}
                        value='extend'
                        checked={this.state.selectedOption === 'extend'}
                        onChange={this.onReview}
                    />
                </TDStyled>
                <TDStyled color={color} align={center}>
                    <RadioButton
                        name={this.props.role + this.props.idx}
                        value='no-action'
                        checked={this.state.selectedOption === 'no-action'}
                        onChange={this.onReview}
                    />
                </TDStyled>
                <TDStyled color={color} align={center}>
                    <RadioButton
                        name={this.props.role + this.props.idx}
                        value='delete'
                        checked={this.state.selectedOption === 'delete'}
                        onChange={this.onReview}
                    />
                </TDStyled>
            </TrStyled>
        );
        return rows;
    }
}
