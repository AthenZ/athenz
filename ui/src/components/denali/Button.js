/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { cx, css } from 'emotion';
import { rgba } from 'polished';
import { colors } from './styles/colors';
import { cssFontFamilies, cssFontWeights } from './styles/fonts';

// NOTE: Psuedo class priority:
// Link Visited Hover Focus Active
// "Lord Vader Hates Furry Animals"

const buttonBase = (props) => css`
    border: none;
    border-radius: 2px;
    box-shadow: 0 0 0 1px transparent inset, 0 0 0 0 ${colors.grey500} inset;
    box-sizing: border-box;
    cursor: pointer;
    display: inline-block;
    ${cssFontFamilies.default};
    ${cssFontWeights.normal};
    font-size: ${props.size === 'large'
        ? '18px'
        : props.size === 'regular'
        ? '14px'
        : '12px'};
    line-height: 1;
    margin: 5px;
    min-width: ${props.size === 'small' ? '70px' : undefined};
    outline: 0;
    padding: ${props.size === 'large'
        ? '12px 28px'
        : props.size === 'regular'
        ? '10px 24px'
        : '8px 18px'};
    text-align: center;
    text-shadow: none;
    text-transform: none;
    transition: ${!props.noanim ? 'all 0.2s ease-in' : undefined};
    vertical-align: baseline;
    white-space: nowrap;
    label: button;
    &:first-of-type {
        margin-left: 0;
    }
    &:disabled {
        background: ${props.dark
            ? rgba(colors.white, 0.1)
            : rgba(colors.grey800, 0.1)};
        color: ${props.dark
            ? rgba(colors.white, 0.2)
            : rgba(colors.grey800, 0.2)};
        cursor: not-allowed;
    }
`;

const buttonVariants = {
    primary: (props) => css`
        background: linear-gradient(
            to right,
            ${props.active ? colors.brand700 : colors.brand600},
            ${props.active ? colors.brand800 : colors.brand700}
        );
        color: ${colors.white};
        &:hover:not(:disabled) {
            background: linear-gradient(
                to right,
                ${colors.brand700},
                ${colors.brand800}
            );
        }
    `,
    secondary: (props) => css`
        background: ${props.dark
            ? props.active
                ? `linear-gradient(to right, ${rgba(
                      colors.brand700,
                      0.3
                  )}, ${rgba(colors.brand800, 0.3)})`
                : 'transparent'
            : props.active
            ? colors.brand200
            : 'transparent'};
        box-shadow: 0 0 0 1px ${colors.brand600} inset;
        color: ${props.dark ? colors.white : colors.brand600};
        &:disabled {
            box-shadow: ${props.dark
                ? `0 0 0 1px ${rgba(colors.white, 0.2)} inset`
                : `0 0 0 1px ${rgba(colors.grey800, 0.1)} inset`};
        }
        &:hover:not(:disabled) {
            background: ${props.dark
                ? `linear-gradient(to right, ${rgba(
                      colors.brand700,
                      0.3
                  )}, ${rgba(colors.brand800, 0.3)})`
                : colors.brand200};
        }
    `,
    danger: () => css`
        background: ${colors.red500};
        color: ${colors.white};
        &:hover:not(:disabled) {
            background: ${colors.red600};
        }
    `,
};

class Button extends React.PureComponent {
    render() {
        const {
            active,
            className,
            danger,
            dark,
            noanim,
            secondary,
            size,
            ...rest
        } = this.props;

        const cssProps = { active, dark, noanim, size };

        const makeCssButton = (props) => css`
            ${buttonBase(props)};
            ${(danger
                ? buttonVariants.danger
                : secondary
                ? buttonVariants.secondary
                : buttonVariants.primary)(props)};
        `;
        const cssButton = makeCssButton(cssProps);

        const classes = cx(cssButton, 'denali-button', className);

        return <button className={classes} {...rest} />;
    }
}

Button.propTypes = {
    /** Set active state */
    active: PropTypes.bool,
    /** Additonal class to apply to the button */
    className: PropTypes.string,
    /** Error color scheme */
    danger: PropTypes.bool,
    /** Dark color scheme (can be used in conjunction with 'secondary') */
    dark: PropTypes.bool,
    /** Disable the button (no clicks) */
    disabled: PropTypes.bool,
    /** Forward React ref to the `<input>` element */
    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    /** Name of the `button` element */
    name: PropTypes.string,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Secondary color scheme */
    secondary: PropTypes.bool,
    /** Size of buttons (default is regular) */
    size: PropTypes.oneOf(['large', 'regular', 'small']),
    /** <button> type attribute */
    type: PropTypes.oneOf(['button', 'submit', 'reset']),
    /** <button> value attribute */
    value: PropTypes.any,
    /** Gets called when the button is clicked */
    onClick: PropTypes.func,
};

Button.defaultProps = {
    active: false,
    danger: false,
    dark: false,
    disabled: false,
    noanim: false,
    size: 'regular',
    secondary: false,
};

export default Button;
