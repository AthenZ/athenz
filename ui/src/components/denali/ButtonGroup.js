/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { css, cx } from '@emotion/css';
import { rgba } from 'polished';
import { colors, cssFontStyles } from './styles/index';

const makeCssButtonGroup = (props) => css`
    align-items: stretch;
    background: ${props.dark
        ? rgba(colors.grey300, 0.1)
        : rgba(colors.brand700, 0.05)};
    border-radius: 2px;
    box-sizing: border-box;
    display: inline-grid;
    grid-template-columns: repeat(${props.length}, 1fr);
    height: 28px;
    position: relative;
    label: button-group;

    /* Active box */
    &::before {
        background: ${colors.white};
        border: 2px solid ${colors.blue500};
        border-radius: 2px;
        box-sizing: border-box;
        content: '';
        height: calc(100% + 4px);
        margin: -2px -2px -2px calc(${props.marginRight}% - 2px);
        position: absolute;
        transition: ${!props.noanim ? 'all 0.3s ease-in' : undefined};
        width: calc(${props.segmentPct}% + 4px);
        label: button-group-active-box;
    }
`;

const makeCssButton = (props) => css`
    align-items: center;
    border-radius: 2px;
    box-sizing: border-box;
    color: ${props.active ? colors.black : colors.brand600};
    cursor: ${props.active ? 'initial' : 'pointer'};
    display: flex;
    ${cssFontStyles.default};
    justify-content: center;
    margin: 3px;
    padding: 0 12px;
    transition: ${!props.noanim ? 'color 0.2s ease-in 0.1s' : undefined};
    white-space: nowrap;
    z-index: 1;
    label: button-group-button;
    &:hover {
        color: ${props.active ? colors.black : colors.brand700};
    }
`;

class ButtonGroup extends React.PureComponent {
    render() {
        const { buttons, className, dark, id, noanim, selectedName } =
            this.props;

        if (!buttons.length) return null;

        const selected = selectedName || buttons[0].name;
        const activeIdx = buttons.findIndex((b) => b.name === selected);

        const renderedButtons = buttons.map((button) => {
            const active = selected === button.name;
            const cssProps = { active, noanim };
            const cssButton = makeCssButton(cssProps);
            return (
                <div
                    className={cssButton}
                    data-active={active}
                    id={button.id ? `button-${button.id}` : undefined}
                    key={button.name}
                    onClick={() => !active && this.props.onClick(button)}
                >
                    {button.label}
                </div>
            );
        });

        const length = this.props.buttons.length;
        const segmentPct = 100 / length;
        const marginRight = activeIdx * segmentPct;
        const cssProps = {
            dark,
            length,
            noanim,
            segmentPct,
            marginRight,
        };

        const cssButtonGroup = makeCssButtonGroup(cssProps);
        const classes = cx(cssButtonGroup, 'denali-button-group', className);

        return (
            <div className={classes} id={id} data-testid='buttongroup'>
                {renderedButtons}
            </div>
        );
    }
}

ButtonGroup.propTypes = {
    /** Array of button configs */
    buttons: PropTypes.arrayOf(
        PropTypes.shape({
            name: PropTypes.string.isRequired,
            label: PropTypes.string.isRequired,
            id: PropTypes.string,
            /** Additional data can be placed here that will be returned on `onClick()` */
            extra: PropTypes.object,
        })
    ).isRequired,
    /** Additonal class to apply to the outer div */
    className: PropTypes.string,
    /** Dark color scheme (can be used in conjunction with `primary`/`secondary`) */
    dark: PropTypes.bool,
    /** id to attach to outer div */
    id: PropTypes.string,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Handler for button click. The clicked button config is returned */
    onClick: PropTypes.func.isRequired,
    /** Specifiy selected button. Defaults to the first button */
    selectedName: PropTypes.string,
};

ButtonGroup.defaultProps = {
    dark: false,
    noanim: false,
};

export default ButtonGroup;
