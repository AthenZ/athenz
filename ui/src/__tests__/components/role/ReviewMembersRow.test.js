/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import ReviewMembersRow from '../../../components/role/ReviewMembersRow';
import { colors } from '../../../components/denali/styles';

describe('ReviewMembersRow', () => {
    it('should render', () => {
        let idx = 'role-review-0';
        let details = {
            memberName: 'user.abhijetv',
            expiration: '2020-04-09T04:43:07.028Z',
            approved: true,
            auditRef: 'Added using Athenz UI',
        };
        let role = 'self-serve1';
        let color = colors.white;
        let submittedReview = false;

        const { getByTestId } = render(
            <table>
                <tbody>
                    <ReviewMembersRow
                        details={details}
                        role={role}
                        color={color}
                        idx={idx}
                        submittedReview={submittedReview}
                    />
                </tbody>
            </table>
        );
        const roleReviewRow = getByTestId('role-review-row');

        expect(roleReviewRow).toMatchSnapshot();
    });
});
