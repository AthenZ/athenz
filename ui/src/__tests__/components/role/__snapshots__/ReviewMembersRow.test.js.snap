// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewMembersRow should render 1`] = `
.emotion-3 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  box-sizing: border-box;
  display: inline-block;
  height: 16px;
  line-height: 16px;
  margin: 3px;
  outline: 0px;
  position: relative;
  vertical-align: bottom;
}

.emotion-3 > input[type='radio'] {
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  position: absolute;
  top: 0;
  visibility: hidden;
}

.emotion-3 > input[type='radio'] + label {
  box-sizing: border-box;
  cursor: pointer;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  height: 16px;
  line-height: 16px;
  padding: 0 10px 0 0;
  vertical-align: bottom;
  white-space: nowrap;
}

.emotion-3 > input[type='radio'] + label:before {
  background: #fff;
  border: 2px solid rgba(54,151,242,0.5);
  border-radius: 50%;
  box-sizing: border-box;
  content: '';
  cursor: pointer;
  display: inline-block;
  height: 16px;
  line-height: 16px;
  margin: 0px 3px -3px 0px;
  -webkit-transition: all 0.25s ease-in;
  transition: all 0.25s ease-in;
  width: 16px;
}

.emotion-3 > input[type='radio']:hover + label:before {
  border: 2px solid #3697f2;
}

.emotion-3 > input[type='radio']:focus + label:before {
  border: 2px solid rgba(54,151,242,0.5);
}

.emotion-3 > input[type='radio']:checked + label:before {
  border: 2px solid #3697f2;
}

.emotion-3 > input[type='radio']:disabled + label {
  color: #d5d5d5;
  cursor: not-allowed;
}

.emotion-3 > input[type='radio']:disabled + label:before {
  border: 2px solid #d5d5d5;
  cursor: not-allowed;
}

.emotion-3 > input[type='radio']:checked + label:after {
  background-color: #3697f2;
  border-radius: 50%;
  box-sizing: border-box;
  content: '';
  height: 8px;
  line-height: 8px;
  left: 4px;
  position: absolute;
  top: 4px;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  width: 8px;
}

.emotion-0 {
  background-color: #fff;
  text-align: left;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
  border-bottom: 1px solid #d5d5d5;
}

.emotion-4 {
  background-color: #fff;
  text-align: center;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
  border-bottom: 1px solid #d5d5d5;
}

<tr
  class="emotion-9"
  data-testid="role-review-row"
>
  <td
    class="emotion-0"
    color="#fff"
  >
    user.abhijetv
  </td>
  <td
    class="emotion-0"
    color="#fff"
  />
  <td
    class="emotion-0"
    color="#fff"
  >
    2020-04-09 04:43 UTC
  </td>
  <td
    class="emotion-4"
    color="#fff"
  >
    <div
      class="emotion-3 denali-radiobutton"
      data-testid="radiobutton-wrapper"
    >
      <input
        checked=""
        id="radiobutton-self-serve1role-review-0-extend"
        name="self-serve1role-review-0"
        type="radio"
        value="extend"
      />
      <label
        for="radiobutton-self-serve1role-review-0-extend"
      />
    </div>
  </td>
  <td
    class="emotion-4"
    color="#fff"
  >
    <div
      class="emotion-3 denali-radiobutton"
      data-testid="radiobutton-wrapper"
    >
      <input
        id="radiobutton-self-serve1role-review-0-no-action"
        name="self-serve1role-review-0"
        type="radio"
        value="no-action"
      />
      <label
        for="radiobutton-self-serve1role-review-0-no-action"
      />
    </div>
  </td>
  <td
    class="emotion-4"
    color="#fff"
  >
    <div
      class="emotion-3 denali-radiobutton"
      data-testid="radiobutton-wrapper"
    >
      <input
        id="radiobutton-self-serve1role-review-0-delete"
        name="self-serve1role-review-0"
        type="radio"
        value="delete"
      />
      <label
        for="radiobutton-self-serve1role-review-0-delete"
      />
    </div>
  </td>
</tr>
`;
